test/flink_test.go

```diff
@@ -120,3 +152,244 @@ func (s *CLITestSuite) TestFlink_Autocomplete() {
+	if err := os.Setenv(timezoneEnvVar, "Europe/London"); err != nil {
+		return envVars, err
+	}
+	envVars = append(envVars, timezoneEnvVar)
+
+	return envVars, nil
+}
+
+func cleanupEnvVars(envVars []string) {
+	for _, envVar := range envVars {
+		os.Unsetenv(envVar)
+	}
+}
+
+func killCommandIfTimeoutExceeded(timeout time.Duration, cmd *exec.Cmd) {
+	if timeout == 0 {
+		timeout = 10 * time.Second
+	}
+	time.Sleep(timeout)
+	if err := cmd.Process.Kill(); err != nil {
+		fmt.Printf("failed to kill command %v/n", err)
+	}
+}
+
+func waitForFlinkShellToBeReady(scanner *bufio.Scanner, output *strings.Builder) {
+	for scanner.Scan() {
+		line := scanner.Text()
+		output.WriteString(line + "\n")
+		if strings.Contains(line, "[Ctrl-Q] Quit [Ctrl-S] Toggle Smart Completion") {
+			break
+		}
+	}
+}
+
+func executeCommands(stdin *os.File, commands []string, stdoutScanner *bufio.Scanner, output *strings.Builder) error {
+	// add exit command to ensure we always close the flink shell
+	commands = append(commands, "exit")
+	for _, command := range commands {
+		// Simulate the user entering a command and add a new line to flush the output buffer
+		_, err := stdin.WriteString(command + "\n")
+		if err != nil {
+			return err
+		}
+
+		for stdoutScanner.Scan() {
+			// Strip all terminal control sequences and skip empty lines
+			line := removeAnsiEscapeSequences(stdoutScanner.Text())
+			if line == "" {
+				continue
+			}
+
+			// Record the output
+			output.WriteString(line + "\n")
+
+			// Once we've seen our command be printed, we know we can continue to submit the statement.
+			// This effectively means we wait for the previous command to complete, since the new command will only
+			// be printed, if the previous one is finished.
+			if strings.Contains(line, fmt.Sprintf("> %s", command)) {
+				err := submitStatement(stdin)
+				if err != nil {
+					return err
+				}
+				break
+			}
+		}
+	}
+	return nil
+}
+
+func removeAnsiEscapeSequences(input string) string {
+	stripColors := regexp.MustCompile(`\x1b\[[0-9;]*[JKmsu]`)
+	stripBellCharacter := regexp.MustCompile(`\a`)
+	stripTitle := regexp.MustCompile(`\x1B]2;`)
+	stripSqlPrompt := regexp.MustCompile(`sql-prompt`)
+	stripErasePrefix := regexp.MustCompile(`> \x1b\[2D`)
+	stripCursorUp := regexp.MustCompile(`\x1b\[1A.*`)
+
+	input = stripColors.ReplaceAllString(input, "")
+	input = stripBellCharacter.ReplaceAllString(input, "")
+	input = stripTitle.ReplaceAllString(input, "")
+	input = stripSqlPrompt.ReplaceAllString(input, "")
+	input = stripErasePrefix.ReplaceAllString(input, "")
+	input = stripCursorUp.ReplaceAllString(input, "")
+
+	return strings.TrimSpace(input)
+}
+
+func submitStatement(stdin *os.File) error {
+	// We need to wait before and after writing a new line, so go-prompt receives the ENTER as an individual key press
+	time.Sleep(50 * time.Millisecond)
+	_, err := stdin.WriteString("\n")
+	time.Sleep(50 * time.Millisecond)
```

Comment:
I see a few `Sleep()` functions hidden throughout these tests. How long do these tests take end-to-end? Can we decrease this to 10ms? Can we scan for specific text to know when we can write a newline?

Reply:
hmm let me see, maybe we can wait for the `Statement successfully submitted.` text 🤔

Reply:
okay that works, added a helper that waits for a particular line to appear and then returns the output recorded up until that point

Reply:
> How long do these tests take end-to-end?

oh and the tests take about 7s to run locally atm:
```
--- PASS: TestCLI (6.73s)
    --- PASS: TestCLI/TestFlinkShell (4.34s)
        --- PASS: TestCLI/TestFlinkShell/use-catalog (1.67s)
        --- PASS: TestCLI/TestFlinkShell/use-database (0.81s)
        --- PASS: TestCLI/TestFlinkShell/set-single-key (0.54s)
        --- PASS: TestCLI/TestFlinkShell/reset-single-key (0.61s)
        --- PASS: TestCLI/TestFlinkShell/reset-all-keys (0.73s)

```

Reply:
Huh, I wonder if that's including the time to build the binary? Let's see where else we can speed up these tests... 1s per integration test is not going to scale well

Reply:
I think we could create another ticket on how to increase the performance here, wdyt @brianstrauch? Yannick has been working on this over a week, and the whole ticket already required much more time than we planned

Reply:
Also, I don't think we're planning to add 100s of integration tests now. However, if we start adding more we'll prioritize increasing the performance first

Reply:
on CI the test takes about 2s, which seems reasonable to me 🤔  
![image](https://github.com/confluentinc/cli/assets/37211050/961fcd4b-a2f3-4a38-b1a4-eb3a3ee8e522)